ospf6d/ospf6_interface.c

```diff
@@ -2801,3 +2859,199 @@ void install_element_ospf6_debug_interface(void)
+{
+	if (CHECK_FLAG(at_cfg->flags, OSPF6_AUTH_TRAILER_KEYCHAIN))
+		vty_out(vty, " ipv6 ospf6 authentication keychain %s\n",
+			at_cfg->keychain);
+	else if (CHECK_FLAG(at_cfg->flags, OSPF6_AUTH_TRAILER_MANUAL_KEY))
+		vty_out(vty,
+			" ipv6 ospf6 authentication key-id %d hash-algo %s key %s\n",
+			at_cfg->key_id,
+			keychain_get_algo_name_by_id(at_cfg->hash_algo),
+			at_cfg->auth_key);
+}
+
+DEFUN(ipv6_ospf6_intf_auth_trailer_keychain,
+      ipv6_ospf6_intf_auth_trailer_keychain_cmd,
+      "ipv6 ospf6 authentication keychain KEYCHAIN_NAME",
+      IP6_STR OSPF6_STR
+      "Enable authentication on this interface\n"
+      "Keychain\n"
+      "Keychain name\n")
+{
+	VTY_DECLVAR_CONTEXT(interface, ifp);
+	int keychain_idx = 4;
+	struct ospf6_interface *oi;
+
+	assert(ifp);
+	oi = (struct ospf6_interface *)ifp->info;
+	if (oi == NULL)
+		oi = ospf6_interface_create(ifp);
+
+	assert(oi);
+	if (CHECK_FLAG(oi->at_cfg.flags, OSPF6_AUTH_TRAILER_MANUAL_KEY)) {
+		vty_out(vty,
+			"Manual key configured, unconfigure it before configuring key chain\n");
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	SET_FLAG(oi->at_cfg.flags, OSPF6_AUTH_TRAILER_KEYCHAIN);
+	if (oi->at_cfg.keychain)
+		XFREE(MTYPE_OSPF6_AUTH_KEYCHAIN, oi->at_cfg.keychain);
+
+	oi->at_cfg.keychain =
+		XSTRDUP(MTYPE_OSPF6_AUTH_KEYCHAIN, argv[keychain_idx]->arg);
+
+	return CMD_SUCCESS;
+}
+
+DEFUN(no_ipv6_ospf6_intf_auth_trailer_keychain,
+      no_ipv6_ospf6_intf_auth_trailer_keychain_cmd,
+      "no ipv6 ospf6 authentication keychain [KEYCHAIN_NAME]",
+      NO_STR IP6_STR OSPF6_STR
+      "Enable authentication on this interface\n"
+      "Keychain\n"
+      "Keychain name\n")
+{
+	VTY_DECLVAR_CONTEXT(interface, ifp);
+	struct ospf6_interface *oi;
+
+	assert(ifp);
+	oi = (struct ospf6_interface *)ifp->info;
+	if (oi == NULL)
+		oi = ospf6_interface_create(ifp);
+
+	assert(oi);
+	if (!CHECK_FLAG(oi->at_cfg.flags, OSPF6_AUTH_TRAILER_KEYCHAIN))
+		return CMD_SUCCESS;
+
+	if (oi->at_cfg.keychain) {
+		oi->at_cfg.flags = 0;
+		XFREE(MTYPE_OSPF6_AUTH_KEYCHAIN, oi->at_cfg.keychain);
+		oi->at_cfg.keychain = NULL;
+	}
+
+	return CMD_SUCCESS;
+}
+
+DEFUN(ipv6_ospf6_intf_auth_trailer_key, ipv6_ospf6_intf_auth_trailer_key_cmd,
+      "ipv6 ospf6 authentication key-id (1-65535) hash-algo "
+#ifdef CRYPTO_INTERNAL
+      "<md5|hmac-sha-256> "
+#endif /* CRYPTO_INTERNAL */
+#ifdef CRYPTO_OPENSSL
+      "<md5|hmac-sha-256|hmac-sha-1|hmac-sha-384|hmac-sha-512> "
+#endif /* CRYPTO_OPENSSL */
+      "key WORD",
+      IP6_STR OSPF6_STR
+      "Authentication\n"
+      "Key ID\n"
+      "Key ID value\n"
+      "Cryptographic-algorithm\n"
+#ifdef CRYPTO_INTERNAL
+      "Use MD5 algorithm\n"
+      "Use HMAC-SHA-256 algorithm\n"
+#endif /* CRYPTO_INTERNAL */
+#ifdef CRYPTO_OPENSSL
+      "Use MD5 algorithm\n"
+      "Use HMAC-SHA-1 algorithm\n"
+      "Use HMAC-SHA-256 algorithm\n"
+      "Use HMAC-SHA-384 algorithm\n"
+      "Use HMAC-SHA-512 algorithm\n"
+#endif /* CRYPTO_OPENSSL */
```

Comment:
`#ifdef` in the middle of `DEFUN` is very tricky and would break e.g. `clippy` (which you're not using here, but probably should too.)

Also, I think it would be better to always accept all types but print errors when they cannot be used.  Otherwise the error for "algorithm not supported" is just "command not found" which is very confusing for an user because they cannot know if they just mistyped the command or FRR was compiled w/o the algo.

(... which makes the `#ifdef` go away too.)

Reply:
I am not aware of clippy. Can you please share more details on the failure scenarios using clippy.

How do you determine run time if you have compiled with --with-crypto=openssl ? 

Intension is to ensure you don't execute command if you have not compiled with openssl.

Reply:
`clippy` is the tool that handles `DEFPY` commands.  It does not support `#ifdef` in the middle of the macro and will error out.  It's not really the main reason here, I just wanted to raise awareness.

> How do you determine run time if you have compiled with --with-crypto=openssl ?

Don't.  Make all options always available.  That's what I meant with "confusing for an user."  If you make the CLI conditional, the user gets "command not found".  It should be "algorithm not supported", you can print that from inside the CLI function.  This is the main reason.

Reply:
Done